[hop] support base_hop._gen_schema

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #150369
* __->__ #149688

This PR creates two utils for generating a schema for hops from example inputs and use base hop as an exmaple.
1. HopArgumentInfoGen creates an argument or an output schema with mutation information.
2. CFuncitonSchemaGen piece together the argument info of inputs and outputs and produces torch._C.FunctionSchema.

is_write attribute of argument info can be computed. Note that the is_write annotation only works when the inputs are flattened (e.g. cannot support mutation inside tuple). We need special handling the case where we have tuple inputs like cond.


cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @chenyang78 @kadeng @chauhang @amjames

ydwu4 committed:
[wip][hop] support base_hop._gen_schema

[ghstack-poisoned]

ydwu4 committed:
Update on "[wip][hop] support base_hop._gen_schema"




cc EikanWang jgong5 wenzhe-nrv sanchitintel voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ydwu4 committed:
Update on "[wip][hop] support base_hop._gen_schema"




cc EikanWang jgong5 wenzhe-nrv sanchitintel voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ydwu4 committed:
Update on "[wip][hop] support base_hop._gen_schema"




cc EikanWang jgong5 wenzhe-nrv sanchitintel voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ydwu4 committed:
Update on "[wip][hop] support base_hop._gen_schema"




cc EikanWang jgong5 wenzhe-nrv sanchitintel voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ydwu4 committed:
Update on "[wip][hop] support base_hop._gen_schema"




cc EikanWang jgong5 wenzhe-nrv sanchitintel voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ydwu4 committed:
Update on "[wip][hop] support base_hop._gen_schema"




cc EikanWang jgong5 wenzhe-nrv sanchitintel voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ydwu4 committed:
Update on "[wip][hop] support base_hop._gen_schema"




cc EikanWang jgong5 wenzhe-nrv sanchitintel voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ydwu4 committed:
Update on "[wip][hop] support base_hop._gen_schema"




cc EikanWang jgong5 wenzhe-nrv sanchitintel voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ydwu4 committed:
Update on "[wip][hop] support base_hop._gen_schema"




cc EikanWang jgong5 wenzhe-nrv sanchitintel voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ydwu4 committed:
Update on "[hop] support base_hop._gen_schema"


This PR creates two utils for generating a schema for hops from example inputs and use base hop as an exmaple.
1. HopArgumentInfoGen creates an argument or an output schema with mutation information.
2. CFuncitonSchemaGen piece together the argument info of inputs and outputs and produces torch._C.FunctionSchema.

is_write attribute of argument info can be computed. Note that the is_write annotation only works when the inputs are flattened (e.g. cannot support mutation inside tuple). We need special handling the case where we have tuple inputs like cond.


cc EikanWang jgong5 wenzhe-nrv sanchitintel voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames 

[ghstack-poisoned]

zou3519 commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -1877,6 +1917,10 @@ void initJitScriptBindings(PyObject* module) {
   m.def("_test_only_remove_upgraders", &test_only_remove_upgraders);
 
+  m.def("_create_argument", &createArgument);
+  m.def("_create_function_schema", &createFunctionSchema);
```
Can you just expose the FunctionSchema constructor to Python instead of adding a `_create_function_schema` function?

zou3519 replied:
Same for create_argument and create_alias_info. and create_tuple_type.

ydwu4 replied:
Sure

zou3519 commented on torch/_ops.py:
```diff
@@ -470,6 +470,18 @@ def wrapper():
         return wrapper()
 
+    # NOTE [HigherOrderOp Schema]
+    # Each invocation of a HigherOrderOperator should have its own schema because
+    # the subgraphs and the arguments can be different even for the same hop.
```
Write somewhere that if this is not implemented the we expect your HOP to be functional (e.g. no mutation of inputs, view aliasing has constraints, etc)

zou3519 commented on torch/_ops.py:
```diff
@@ -470,6 +470,18 @@ def wrapper():
+    # mutates its inputs or have output aliasing input.
+    def gen_schema(self, *args, **kwargs):
+        raise NotImplementedError(
+            f"HigherOrderOp {self._name} does not implement a gen_schema."
```
nit: spell out Operator, I know we don't in a lot of places but it's nicer to read

zou3519 commented on aten/src/ATen/core/function_schema.cpp:
```diff
@@ -250,6 +250,7 @@ std::ostream& operator<<(std::ostream& out, const FunctionSchema& schema) {
     return_ss << returns.at(0);
     auto return_str = return_ss.str();
 
+    // Note [schema for signle return item with parenthesis]
```
nit: typo in signle

zou3519 commented on test/dynamo/test_base_hop.py:
```diff
@@ -73,6 +74,88 @@ def forward(self, l_x_: "f32[3, 3]", l_y_: "f32[3, 3]"):
+                _get_example_value,
+                (node.args, node.kwargs),
+            )
+            schema = node.target.gen_schema(*fake_args, **fake_kwargs)
```
Is there a way to include the word "imputed" somehow in the gen_schema API? gen_imputed_schema? I've been using "Imputed schema for HOPs" to refer to this project

zou3519 replied:
Hmm maybe gen_schema is better, it's shorter

zou3519 commented on test/dynamo/test_base_hop.py:
```diff
@@ -73,6 +74,88 @@ def forward(self, l_x_: "f32[3, 3]", l_y_: "f32[3, 3]"):
+        self.assertEqual(len(schemas), 1)
+        self.assertExpectedInline(
+            str(schemas[0]),
+            # See Note [schema for signle return item with parenthesis] for why
+            # there's an extra parenthesis in schema of return
+            """invoke_quant_test(Any subgraph, Tensor(a1!) arg0, Tensor(a2!) arg1, str scheme="nf4") -> ((Tensor) out)""",  # noqa: B950
```
In aten operators, the output is not actually named. Can we avoid naming the output?

When the output of aten operators is named we actually return a named tuple, which is not the case for HOPs

zou3519 replied:
I think this will solve your multiple set of parenthesis problem too. I expect the schema to look like:
```
invoke_quant_test(Any subgraph, Tensor(a1!) arg0, Tensor(a2!) arg1, str scheme="nf4") -> Tensor
```

ydwu4 replied:
For the return of hop schema, we  have two options: 

1. we could return **"->  Tensor", and "-> Tensor, Tensor"** when return is a tuple of 1 and 2 by constructing the FunctionArugment with returns = [*tuple_output] but we have no way to differentiate tuple of size 1 and a single tensor: they're both **"-> Tensor"** (although for hops we should always return a tuple? there's no confusion if enforce this invariant). 

2. Or we set the returns to be [tuple_output], we'll get **"((Tensor))"** or **"((Tensor, Tensor))"** . In this case, there's no confusion but we have the double bracket.

zou3519 commented on test/dynamo/test_base_hop.py:
```diff
@@ -73,6 +74,88 @@ def forward(self, l_x_: "f32[3, 3]", l_y_: "f32[3, 3]"):
+            str(schemas[0]),
+            # See Note [schema for signle return item with parenthesis] for why
+            # there's an extra parenthesis in schema of return
+            """invoke_quant_test(Any subgraph, Tensor(a1!) arg0, Tensor arg1, str scheme="nf4") -> ((Tensor, Tensor, Tensor, Tensor) out)""",  # noqa: B950
```
I expected this to look like:
```
invoke_quant_test(Any subgraph, Tensor(a1!) arg0, Tensor arg1, str scheme="nf4") -> (Tensor, Tensor, Tensor, Tensor)
```

zou3519 commented on test/dynamo/test_base_hop.py:
```diff
@@ -164,11 +247,8 @@ def f(inner, x, y):
-            "Encountered input mutation during higher order op tracing for HOP",
-        ):
-            f(inner2, x, y)
+        # input mutation can be supported with auto_functionalize
```
Why did this change?

ydwu4 replied:
We need the change to test the input mutation schema generation. Can also delete the input mutation test and defer it to next PR where we add support.

zou3519 reviewed (commented):
got through half of this will be back to review the rest later

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -3080,7 +3080,7 @@ def hash_graph_and_inputs(tx, gmod, fake_inputs):
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
-        self.supports_input_mutation = False
+        self.supports_input_mutation = True
```
not all of these support input mutation?

ydwu4 replied:
yeah, it's just for testing. Let me refactor it out.

zou3519 commented on torch/_library/infer_schema.py:
```diff
@@ -318,3 +319,149 @@ def tuple_to_list(tuple_type: type[tuple]) -> type[list]:
         return list[typing.Union[tuple(type_args)]]  # type: ignore[misc, return-value]
+
+
+# Below is an implementation of generating FunctionSchema from example values.
```
Can you toss this into a new file? infer_schema is for inferring the schema from a function. I think you just want a hop_schema file.

Maybe even toss it under torch._higher_order_ops.schema

ydwu4 replied:
Sure, torch._higher_order_ops.schema sounds good.

zou3519 commented on torch/_library/infer_schema.py:
```diff
@@ -318,3 +319,149 @@ def tuple_to_list(tuple_type: type[tuple]) -> type[list]:
+    def from_example(obj: Any) -> Any:
+        import torch
+
+        if isinstance(obj, torch.Tensor):
+            return torch._C.TensorType.get()
+        elif isinstance(obj, torch.fx.GraphModule):
+            return torch._C.AnyType.get()
+        elif isinstance(obj, torch.SymInt):
+            return (torch._C.SymIntType.get(),)
+        elif isinstance(obj, torch.SymBool):
+            return (torch._C.SymBoolType.get(),)
+        elif isinstance(obj, tuple):
+            return torch._C._create_tuple_type(
+                tuple(CTypeGen.from_example(arg) for arg in obj)
+            )
+        elif obj is None:
+            return torch._C.NoneType.get()
+        tp = type(obj)
+        if tp not in CTypeGen.convert_to_base_ty:
+            raise RuntimeError(f"unsupported type {tp}")
+        return CTypeGen.convert_to_base_ty[tp]
```
I feel like there has to be a C++ function somewhere for this?
Can we try to parse an IValue out of the obj and then use the IValue's type?

And then if it fails, then we can do special cases?

ydwu4 replied:
yeah, let me try to find it.

zou3519 reviewed (commented):
overall looks good, had some comments

ydwu4 committed:
Update on "[hop] support base_hop._gen_schema"


This PR creates two utils for generating a schema for hops from example inputs and use base hop as an exmaple.
1. HopArgumentInfoGen creates an argument or an output schema with mutation information.
2. CFuncitonSchemaGen piece together the argument info of inputs and outputs and produces torch._C.FunctionSchema.

is_write attribute of argument info can be computed. Note that the is_write annotation only works when the inputs are flattened (e.g. cannot support mutation inside tuple). We need special handling the case where we have tuple inputs like cond.


cc EikanWang jgong5 wenzhe-nrv sanchitintel voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames 

[ghstack-poisoned]

ydwu4 committed:
Update on "[hop] support base_hop._gen_schema"


This PR creates two utils for generating a schema for hops from example inputs and use base hop as an exmaple.
1. HopArgumentInfoGen creates an argument or an output schema with mutation information.
2. CFuncitonSchemaGen piece together the argument info of inputs and outputs and produces torch._C.FunctionSchema.

is_write attribute of argument info can be computed. Note that the is_write annotation only works when the inputs are flattened (e.g. cannot support mutation inside tuple). We need special handling the case where we have tuple inputs like cond.


cc EikanWang jgong5 wenzhe-nrv sanchitintel voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames 

[ghstack-poisoned]

ydwu4 commented on torch/_higher_order_ops/schema.py:
```diff
@@ -0,0 +1,154 @@
+            not out_argument_info.is_mutated
+        ), "out_argument_info.is_mutated should always be set to False."
+        rets = None
+        if len(out_argument_info.example_value) == 1:
```
Here is the trick that turns a tuple with single element into "->((Tensor))". If we don't do this, we'll get -> Tensor, which is ambiguous.

zou3519 commented on torch/_ops.py:
```diff
@@ -470,6 +470,23 @@ def wrapper():
+    # are no input, output aliasing via views or direct referencing.
+    def gen_schema(self, *args, **kwargs):
+        raise NotImplementedError(
+            f"HigherOrderOperator {self._name} does not implement a gen_schema."
```
In the error message, say something about that we expect it to be functional

zou3519 commented on test/dynamo/test_base_hop.py:
```diff
@@ -73,6 +74,140 @@ def forward(self, l_x_: "f32[3, 3]", l_y_: "f32[3, 3]"):
+        return schemas
+
+    @torch._dynamo.config.patch(assume_static_by_default=True)
+    def test_schema_gen_single_return(self):
```
Do you have a test with:
1) a None input?
2) an int input?

I would expect those to be treated as constants and not shown in the graph but just want to check

ydwu4 replied:
Yeah, added tests to show this.

ydwu4 committed:
Update on "[hop] support base_hop._gen_schema"


This PR creates two utils for generating a schema for hops from example inputs and use base hop as an exmaple.
1. HopArgumentInfoGen creates an argument or an output schema with mutation information.
2. CFuncitonSchemaGen piece together the argument info of inputs and outputs and produces torch._C.FunctionSchema.

is_write attribute of argument info can be computed. Note that the is_write annotation only works when the inputs are flattened (e.g. cannot support mutation inside tuple). We need special handling the case where we have tuple inputs like cond.


cc EikanWang jgong5 wenzhe-nrv sanchitintel voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames 

[ghstack-poisoned]

ydwu4 committed:
Update on "[hop] support base_hop._gen_schema"


This PR creates two utils for generating a schema for hops from example inputs and use base hop as an exmaple.
1. HopArgumentInfoGen creates an argument or an output schema with mutation information.
2. CFuncitonSchemaGen piece together the argument info of inputs and outputs and produces torch._C.FunctionSchema.

is_write attribute of argument info can be computed. Note that the is_write annotation only works when the inputs are flattened (e.g. cannot support mutation inside tuple). We need special handling the case where we have tuple inputs like cond.


cc EikanWang jgong5 wenzhe-nrv sanchitintel voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames 

[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge

ydwu4 commented:
@pytorchbot merge
